src/slashdbclient.js

```diff
@@ -1,44 +1,79 @@
 const SDB_SDBC_INVALID_USERNAME = 'Invalid username parameter, must be string';
 const SDB_SDBC_INVALID_APIKEY = 'Invalid apiKey parameter, must be string';
 const SDB_SDBC_INVALID_PASSWORD = 'Invalid password parameter, must be string';
-const SDB_SDBC_INVALID_USERNAME_MISMATCH = 'Login username must match object username';
+const SDB_SDBC_INVALID_IDPID = 'Invalid identity provider parameter, must be string';
+const SDB_SDBC_INVALID_REDIRECT_URI = 'Invalid redirect uri parameter, must be string';
+const SDB_SDBC_IDENTITY_PROVIDER_NOT_AVAILABLE = "Identity provider not available in settings";
 
 /** 
  * Stores parameters necessary to communicate with a SlashDB instance and provides methods for retrieving metadata from the instance.
  */
 class SlashDBClient {
 
   /** 
-   * Creates a SlashDB client to connect to a SlashDB instance. 
-   * @param {string} host - hostname/IP address of the SlashDB instance, including protocol and port number (e.g. http://192.168.1.1:8080)
-   * @param {string} [username] - optional username to use when connecting to SlashDB instance
-   * @param {string} [apiKey] - optional API key associated with username
+   * Creates a SlashDB client to connect to a SlashDB instance.
+   * 
+   * @param {Object} config
+   * @param {string} config.host - hostname/IP address of the SlashDB instance, including protocol and port number (e.g. http://192.168.1.1:8080)
+   * @param {string} config.apiKey - optional API key associated with username
+   * @param {Object} config.sso - optional settings to login with Single Sign-On
+   * @param {string} config.sso.idpId - optional identity provider id configured in SlashDB
+   * @param {string} config.sso.redirectUri - optional redirect uri to redirect browser after sign in
+   * @param {boolean} config.sso.popUp - optional flag to sign in against the identity provider with a Pop Up window (false by default)
    */
 
-  constructor(host, username, apiKey) {
+  constructor(config) {
+
+    const host = config.host;
 
     if (!host || typeof(host) !== 'string') {
       throw TypeError(SDB_SDBC_INVALID_HOSTNAME);
-
     }
 
-    if (username && typeof(username) !== 'string') {
-      throw TypeError(SDB_SDBC_INVALID_USERNAME);
+    this.host = host;
 
+    this.username = null;
+    this.apiKey = null;
+    this.basic = null;
+    this.sso = {
+      idpId: null,
+      redirectUri: null,
+      popUp: false
     }
 
-    if (apiKey && typeof(apiKey) !== 'string') {
-      throw TypeError(SDB_SDBC_INVALID_APIKEY);
+    if (config.hasOwnProperty('apiKey')) {
+      const apiKey = config.apiKey;
+      if (apiKey && typeof(apiKey) !== 'string') {
+        throw TypeError(SDB_SDBC_INVALID_APIKEY);
+      }
+      this.apiKey = apiKey;
+    } else if (config.hasOwnProperty('sso')) {
+      const idpId = config.sso.idpId;
+      const redirectUri = config.sso.redirectUri;
+
+      if (idpId && typeof(idpId) !== 'string') {
```

Comment:
what happens if `idpId` or `redirectUri` aren't set?  should that be allowed?

Reply:
Good question, in that case login method will not do anything and will return false, without any warning

Reply:
should probably raise an exception right?  i don't think you can have empty `idpId`.  The `redirectUri` I'm not sure, that might be able to be set empty depending on how you configure the app?